design/KEPS/namespace-discovery/proposal.md

```diff
@@ -12,19 +12,162 @@ Provide a better way for namespace discovery within both server and operator.
+
+## Discovery Selector Kiali Config Option
+
+A new configuration option will be added that will mimic the Istio `discoverySelectors` configuration option. Using the configuration option, users will be able to similarly scope Kiali to the defined set of namespaces. This configuration option would take precedence over the autodiscovered controlplane discovery selectors. Having a config option within Kiali has the advantage of not relying on auto-discovery via the Istio control plane. Kiali does not need to fail if it cannot, even temporarily, access the control plane configuration, because it can rely on its own discovery selectors configuration. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.
+
+Kiali would add a `deployment.discovery_selectors` config option where you can specify global discovery selectors or cluster specific overrides `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to. Cluster specific overrides will take precedence over the global option for that cluster.
+
+```yaml
+discoverySelectors:
+  global:
+    - labelSelector:
+  cluster_name:
+    - labelSelector:
+```
+
+For multi-primary with different discovery selectors:
+
+```yaml
+discoverySelectors:
+  east:
+    - matchLabels:
+        team: backend
+  west:
+    - matchLabels:
+        team: api
+```
+
+## No cluster wide access
+
+Kiali relies on `deployment.accessible_namespaces` when `deployment.cluster_wide_access = false` to know which namespaces the server should create a kube cache for. The operator also uses `deployment.accessible_namespaces` to create a `Role/RoleBinding` for the Kiali server's Service Account giving the Kiali server access to that namespace. In order for the operator to create RBAC rules for the Kiali server, the operator will need to reconcile the discovery selectors with namespaces that exist in the cluster. The operator can take what is set in `deployment.discovery_selectors` and set the Kiali server's `deployment.discovery_selectors` config field to be the result of this reconcilation. For example:
+
+```yaml
+spec:
+  deployment:
+    cluster_wide_access: false
+    discovery_selectors:
+      - matchLabels:
+          team: backend
+```
+
+would be translated into the Kiali config as:
+
+```yaml
+spec:
+  deployment:
+    cluster_wide_access: false
+    discovery_selectors:
+      - matchLabels:
+          kubernetes.io/metadata.name: backend-app1
+      - matchLabels:
+          kubernetes.io/metadata.name: backend-app2
```

Comment:
I actually like this better than the "keep accessible_namespaces" idea - even though the latter would be easier to implement than the former :) We'd completely get rid of accessible_namespaces which avoids any confusion about whether the "old" accessible_namespaces still works or not and just keeps one way in the server to get the list of namespaces - through discovery selectors (and we already have code in the server that does that).